game/scripts/vscripts/items/azazel_tower_defense.lua

```diff
@@ -0,0 +1,158 @@
+LinkLuaModifier("modifier_defense_tower_construction", "items/azazel_tower_defense.lua", LUA_MODIFIER_MOTION_NONE)
+LinkLuaModifier("modifier_defense_tower_true_sight", "items/azazel_tower_defense.lua", LUA_MODIFIER_MOTION_NONE)
+
+item_azazel_tower_defense_1 = class(ItemBaseClass)
+
+function item_azazel_tower_defense_1:CastFilterResultLocation(location)
+  if IsClient() then
+    return UF_SUCCESS -- the client can't use the GridNav, but the server will correct it anyway, you can't cheat that.
+  end
+  if (not GridNav:IsTraversable(location)) or #FindUnitsInRadius(DOTA_TEAM_NEUTRALS, location, nil, 144, DOTA_UNIT_TARGET_TEAM_BOTH, DOTA_UNIT_TARGET_BUILDING, DOTA_UNIT_TARGET_FLAG_INVULNERABLE, FIND_ANY_ORDER, false) > 0 or
+    self:GetCaster():IsPositionInRange(location, 144 + self:GetCaster():GetHullRadius())
+  then
+    return UF_FAIL_CUSTOM
+  else
+    return UF_SUCCESS
+  end
+end
+function item_azazel_tower_defense_1:GetCustomCastErrorLocation(location)
+  return "#dota_hud_error_no_buildings_here"
+end
+
+function item_azazel_tower_defense_1:OnSpellStart()
+  local caster = self:GetCaster()
+  local location = self:GetCursorPosition()
+  local building = CreateUnitByName("npc_azazel_tower_defense", location, true, caster, caster:GetOwner(), caster:GetTeam())
+  building:SetOwner(caster)
+  GridNav:DestroyTreesAroundPoint(location, building:GetHullRadius(), true)
+  building:SetOrigin(location)
+  building:RemoveModifierByName('modifier_invulnerable')
+  building:AddNewModifier(building, self, "modifier_defense_tower_construction", {duration = -1})
+  building:AddNewModifier(building, self, "modifier_defense_tower_true_sight", {duration = -1})
+  local charges = self:GetCurrentCharges() - 1
+  if charges < 1 then
+    caster:RemoveItem(self)
+  else
+    self:SetCurrentCharges(charges)
+  end
+end
+
+-- upgrades
+item_azazel_tower_defense_2 = item_azazel_tower_defense_1
+item_azazel_tower_defense_3 = item_azazel_tower_defense_1
+item_azazel_tower_defense_4 = item_azazel_tower_defense_1
+
+--------------------------------------------------------------------------
+-- base modifier
+
+modifier_defense_tower_construction = class(ModifierBaseClass)
+
+local SINK_HEIGHT = 300
+local THINK_INTERVAL = 0.1
+
+function modifier_defense_tower_construction:OnCreated()
+  if IsServer() then
+    local target = self:GetParent()
+    local ab = self:GetAbility()
+    local maxhealth = target:GetMaxHealth() + ab:GetSpecialValueFor("bonus_health")
+    local location = target:GetOrigin()
+    local time = ab:GetSpecialValueFor("construction_time")
+    target:Attribute_SetIntValue("construction_time", time)
+    target:Attribute_SetIntValue("bonus_damage", ab:GetSpecialValueFor("bonus_damage"))
```

Comment:
I think just saving these as properties on `self` makes more sense, but it's fine as is.